src/create_isurf.cpp

```diff
@@ -0,0 +1,1636 @@
+  svalues = NULL;
+  mvalues = NULL;
+  ivalues = NULL;
+
+  // for communication
+
+  ixyz = NULL;
+  cghost = NULL;
+  sghost = NULL;
+  ighost = NULL;
+  numsend = NULL;
+  maxgrid = maxghost = 0;
+
+  proclist = NULL;
+  locallist = NULL;
+  maxsend = 0;
+
+  sbuf = NULL;
+  maxsbuf = 0;
+}
+
+/* ---------------------------------------------------------------------- */
+
+CreateISurf::~CreateISurf()
+{
+  memory->destroy(cvalues);
+  memory->destroy(svalues);
+  memory->destroy(mvalues);
+  memory->destroy(ivalues);
+
+  memory->destroy(ixyz);
+  memory->destroy(cghost);
+  memory->destroy(sghost);
+  memory->destroy(ighost);
+  memory->destroy(numsend);
+
+  memory->destroy(proclist);
+  memory->destroy(locallist);
+
+  memory->destroy(sbuf);
+}
+
+/* ---------------------------------------------------------------------- */
+
+void CreateISurf::command(int narg, char **arg)
+{
+  if (!grid->exist)
+    error->all(FLERR,"Cannot create_isurf before grid is defined");
+  if (!surf->exist)
+    error->all(FLERR,"Must read in surface first with read_surf");
+  if (surf->implicit && surf->exist)
+    error->all(FLERR,"Cannot have pre-existing implicit surfaces");
+  if (!surf->distributed)
+    error->all(FLERR,"Explicit surface must be distributed");
+
+  if (narg != 4) error->all(FLERR,"Illegal create_isurf command");
+
+  // grid group
+
+  ggroup = grid->find_group(arg[0]);
+  if (ggroup < 0) error->all(FLERR,"Create_isurf grid group ID does not exist");
+  groupbit = grid->bitmask[ggroup];
+
+  // ablate fix
+
+  char *ablateID = arg[1];
+  int ifix = modify->find_fix(ablateID);
+  if (ifix < 0)
+    error->all(FLERR,"Fix ID for read_surf does not exist");
+  if (strcmp(modify->fix[ifix]->style,"ablate") != 0)
+    error->all(FLERR,"Fix for read_surf is not a fix ablate");
+  ablate = (FixAblate *) modify->fix[ifix];
+  if (ggroup != ablate->igroup)
+    error->all(FLERR,"Create_isurf group does not match fix ablate group");
+
+  // threshold for corner value
+
+  thresh = input->numeric(FLERR,arg[2]);
+  if (thresh < 0 || thresh > 255)
+    error->all(FLERR,"Create_isurf thresh must be bounded as (0,255)");
+
+  // mode to determine corner values
+
+  if (strcmp(arg[3],"inout") == 0) aveFlag = 0;
+  else if (strcmp(arg[3],"ave") == 0) aveFlag = 1;
+  else error->all(FLERR,"Unknown surface corner mode called");
+
+  //if (aveFlag && comm->nprocs > 1)
+  //  error->all(FLERR,"Create_isurf averaging not possible in parallel");
+
+  // nxyz already takes into account subcells
+  // find corner values for all grid cells initially
+  // only store those within ggroup when calling ablate->store
+  // 0 check uniform grid for entire domain
+
+  grid->check_uniform_group(ggroup,nxyz,corner,xyzsize);
+  nglocal = grid->nlocal;
+
+  // find all corner values
+
+  set_corners();
+  MPI_Barrier(world);
+
+  // remove old explicit surfaces
+
+  remove_old();
+
+  surf->implicit = 1;
+  surf->exist = 1;
+
+  tvalues = NULL; // TODO: Add per-surface type
```

Comment:
TODO: should this done before merging?

Reply:
Right now, all surface elements are assigned to the same surface type. Assigning per-grid-cell surface types was not a capability I was planning to implement right now. Maybe in the future if it's requested?